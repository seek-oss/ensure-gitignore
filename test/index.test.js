const fs = require('fs');
const path = require('path');
const managedIgnore = require('../index');

const read = filepath =>
  fs.readFileSync(filepath, 'utf-8', err => {
    if (err) {
      throw err;
    }
  });

const unindent = str =>
  (typeof str === 'string' ? [str] : str.raw)
    .join('')
    .split(/\r?\n/)
    .map(s => s.trim())
    .join('\n');

describe('managed-ignore', () => {
  describe('file system test', () => {
    const filepath = path.join(__dirname, 'output/write');

    beforeAll(() => fs.writeFileSync(filepath, '', 'utf-8'));
    afterAll(() => fs.unlinkSync(filepath));

    it('write', () => {
      managedIgnore({
        ignores: ['a'],
        filepath
      });
      expect(read(filepath)).toEqual(unindent`
      a # managed
      `);
    });
  });

  it('empty managed', () => {
    expect(
      managedIgnore({
        ignores: [],
        filepath: path.join(__dirname, 'output/append'),
        dryRun: true
      })
    ).toMatchSnapshot();
  });

  it('nothing managed', () => {
    expect(
      managedIgnore({
        filepath: path.join(__dirname, 'output/append'),
        dryRun: true
      })
    ).toMatchSnapshot();
  });

  it('append', () => {
    expect(
      managedIgnore({
        ignores: ['e'],
        filepath: path.join(__dirname, 'output/append'),
        dryRun: true
      })
    ).toMatchSnapshot();
  });

  it('append preserve whitespace', () => {
    expect(
      managedIgnore({
        ignores: ['e'],
        filepath: path.join(__dirname, 'output/appendPreserveWhitespace'),
        dryRun: true
      })
    ).toMatchSnapshot();
  });

  it('append custom suffix', () => {
    expect(
      managedIgnore({
        ignores: ['e'],
        filepath: path.join(__dirname, 'output/append'),
        suffix: 'managed externally',
        dryRun: true
      })
    ).toMatchSnapshot();
  });

  it('take over ignore', () => {
    expect(
      managedIgnore({
        ignores: ['a/**'],
        filepath: path.join(__dirname, 'output/append'),
        dryRun: true
      })
    ).toMatchSnapshot();
  });

  it('take over ignore exact only', () => {
    expect(
      managedIgnore({
        ignores: ['a'],
        filepath: path.join(__dirname, 'output/append'),
        dryRun: true
      })
    ).toMatchSnapshot();
  });

  it('take over and append new', () => {
    expect(
      managedIgnore({
        ignores: ['b', 'f'],
        filepath: path.join(__dirname, 'output/append'),
        dryRun: true
      })
    ).toMatchSnapshot();
  });

  it('error if file not found', () => {
    const manage = () =>
      managedIgnore({
        filepath: 'output/notfound'
      });

    expect(manage).toThrowErrorMatchingSnapshot();
  });
});
