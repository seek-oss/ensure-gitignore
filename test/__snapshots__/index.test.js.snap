// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`managed-ignore append 1`] = `
"a/**
b
c
d

e # managed
"
`;

exports[`managed-ignore append custom suffix 1`] = `
"a/**
b
c
d

e # managed externally
"
`;

exports[`managed-ignore append preserve whitespace 1`] = `
"a/**
b


c
d

e # managed
"
`;

exports[`managed-ignore empty managed 1`] = `
"a/**
b
c
d
"
`;

exports[`managed-ignore error if file not found 1`] = `"ENOENT: no such file or directory, open 'output/notfound'"`;

exports[`managed-ignore nothing managed 1`] = `
"a/**
b
c
d
"
`;

exports[`managed-ignore take over and append new 1`] = `
"a/**
b # managed
c
d

f # managed
"
`;

exports[`managed-ignore take over ignore 1`] = `
"a/** # managed
b
c
d
"
`;

exports[`managed-ignore take over ignore exact only 1`] = `
"a/**
b
c
d

a # managed
"
`;
